File: python/wiringpi_smp03.py
import wiringpi
from wiringpi import GPIO
import time

# PIN_PWM
PIN_PWM2 = 22

def main():
    # initialize
    wiringpi.wiringPiSetup()
    # gpio mode <pin> pwm
    wiringpi.pinMode(PIN_PWM2, GPIO.PWM_OUTPUT)
    # gpio pwmr <pin> <range>
    wiringpi.pwmSetRange(PIN_PWM2, 1024)
    # gpio pwmc <pin> <clock>
    wiringpi.pwmSetClock(PIN_PWM2, 1)
    # gpio pwm <pin> <value>
    wiringpi.pwmWrite(PIN_PWM2, 512)
    time.sleep(1)
    while(True):
        for brightness in range(4, 10+1):
            wiringpi.pwmWrite(PIN_PWM2, 1 << brightness)
            time.sleep(1)
        wiringpi.pwmWrite(PIN_PWM2, 0)
        time.sleep(1)
        for brightness in range(0, 6+1):
            wiringpi.pwmWrite(PIN_PWM2, 1024 >> brightness)
            time.sleep(1)
        wiringpi.pwmWrite(PIN_PWM2, 1024)
        time.sleep(1)

if __name__ == '__main__':
    main()

File: c_cpp/wiringpi_smp03/wiringpi_smp03.c
#include <wiringPi.h>
#include <stdio.h>

typedef unsigned char uint8;
typedef unsigned short uint16;

/* PIN_PWM */
#define PIN_PWM2 22

int main (void) {
    uint16 brightness;
    /* initialize */
    wiringPiSetup();
    /* gpio mode <pin> pwm */
    pinMode(PIN_PWM2, PWM_OUTPUT);
    /* gpio pwmr <pin> <range> */
    pwmSetRange(PIN_PWM2, 1024);
    /* gpio pwmc <pin> <clock> */
    pwmSetClock(PIN_PWM2, 1);
    /* gpio pwm <pin> <value> */
    pwmWrite(PIN_PWM2, 512);
    delay(1000);
    while(TRUE) {
        for(brightness=4; brightness<=10; brightness++) {
            pwmWrite(PIN_PWM2, 1 << brightness);
            delay(1000);
        }
        pwmWrite(PIN_PWM2, 0);
        delay(1000);
        for(brightness=0; brightness<=6; brightness++) {
            pwmWrite(PIN_PWM2, 1024 >> brightness);
            delay(1000);
        }
        pwmWrite(PIN_PWM2, 1024);
        delay(1000);
    }

    return 0;
}
